kafka-client/src/test/java/dev/responsive/kafka/async/AsyncProcessorIntegrationTest.java

```diff
@@ -0,0 +1,293 @@
+      finalOutputRecords.put(key, finalValue);
+    }
+
+    final Map<String, Object> properties = getMutableProperties();
+    final KafkaProducer<String, String> producer = new KafkaProducer<>(properties);
+    
+    final AtomicInteger processed = new AtomicInteger(0);
+    final Map<String, String> latestValues = new ConcurrentHashMap<>();
+
+    final StreamsBuilder builder = new StreamsBuilder();
+
+    final KStream<String, String> input = builder.stream(inputTopic());
+    input
+        .processValues(
+            createAsyncProcessorSupplier(
+                new UserFixedKeyProcessorSupplier(processed, latestValues)),
+            Named.as("AsyncProcessor"),
+            ASYNC_KV_STORE)
+        .to(outputTopic());
+
+    // The total number of records processed, equal to the total number of output records
+    // ONLY when caching is disabled
+    final int numProcessedRecords = keys.size() * 5;
+
+    try (final var streams = new ResponsiveKafkaStreams(builder.build(), properties)) {
+      startAppAndAwaitRunning(Duration.ofSeconds(10), streams);
+
+      // When:
+      pipeRecords(producer, inputTopic(), inputRecords);
+
+      // Then:
+      final var kvs = readOutput(outputTopic(), 0, numProcessedRecords, true, properties);
+
+      for (final String key : keys) {
+        final String finalValue = finalOutputRecords.get(key);
+        assertThat(kvs, hasItem(new KeyValue<>(key, finalValue)));
+      }
+    }
+    assertThat(processed.get(), equalTo(numProcessedRecords));
+    assertThat(latestValues, equalTo(finalOutputRecords));
+  }
+  
+  private static class UserFixedKeyProcessor implements FixedKeyProcessor<String, String, String> {
+    
+    private final AtomicInteger processed;
+    private final Map<String, String> latestValues;
+    
+    private final String streamThreadName;
+    
+    private int partition;
+    private FixedKeyProcessorContext<String, String> context;
+    private KeyValueStore<String, String> kvStore;
+
+    public UserFixedKeyProcessor(
+        final AtomicInteger processed,
+        final Map<String, String> latestValues
+    ) {
+      this.processed = processed;
+      this.latestValues = latestValues;
+      this.streamThreadName = Thread.currentThread().getName();
+    }
+
+    @Override
+    public void init(final FixedKeyProcessorContext<String, String> context) {
+      this.context = context;
+      this.kvStore = context.getStateStore(ASYNC_KV_STORE);
+      this.partition = context.taskId().partition();
+
+      System.out.printf("stream-thread [%s][%s] Initialized processor%n",
+                        streamThreadName, partition
+      );
+    }
+
+    @Override
+    public void process(final FixedKeyRecord<String, String> record) {
+      System.out.printf("stream-thread [%s][%d] Processing record: <%s, %s>%n",
+                        streamThreadName, partition, record.key(), record.value()
+      );
+      
```

Comment:
did you try instrumenting this with `sleep` calls and making sure the test still passes?

Reply:
it would also be good to run the test with a really low commit interval (longer than the sleep) to make sure the flush path gets exercised when there are queued up records.

Reply:
Ah this was actually some leftover debugging logging, but tbh it might be kind of useful so I'm just going to leave it in. I also configured the tests to run at TRACE for the async processing stuff only, so we have full logs in case of an actual flaky failure

I haven't tried with a sleep but I'll put that on the testing TODO list for sure